File: C#/Day2.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AdventSolutions
{
    /// <summary>
    /// Day 2 solution: Checking if reports are safe or can be made safe by removing one level.
    /// </summary>
    public static class Day2
    {
        /// <summary>
        /// Checks if a report is safe: differences between consecutive elements are all in [-3, -1] or [1, 3].
        /// </summary>
        private static bool IsSafeReport(List<int> report)
        {
            if (report.Count < 2) return true; // Trivially safe if not enough data.

            var differences = report.Zip(report.Skip(1), (a, b) => b - a);

            bool allDecreasing = differences.All(d => d >= -3 && d <= -1);
            bool allIncreasing = differences.All(d => d >= 1 && d <= 3);

            return allDecreasing || allIncreasing;
        }

        /// <summary>
        /// Checks if a report can be made safe by removing zero or one element.
        /// </summary>
        private static bool IsSafeWithDampener(List<int> report)
        {
            if (IsSafeReport(report)) return true;

            for (int i = 0; i < report.Count; i++)
            {
                var modified = new List<int>(report);
                modified.RemoveAt(i);
                if (IsSafeReport(modified))
                    return true;
            }

            return false;
        }

        /// <summary>
        /// Count the number of safe reports from a file, with optional dampener logic.
        /// </summary>
        private static int CountSafeReports(string filename, bool withDampener = false)
        {
            int safeCount = 0;
            foreach (var line in File.ReadLines(filename))
            {
                var report = line.Split(' ', StringSplitOptions.RemoveEmptyEntries)
                                 .Select(int.Parse)
                                 .ToList();
                bool safe = withDampener ? IsSafeWithDampener(report) : IsSafeReport(report);
                if (safe) safeCount++;
            }
            return safeCount;
        }

        public static void Run()
        {
            string filePath = "../data/Day2.txt";
            if (!File.Exists(filePath))
            {
                Console.WriteLine("Day2 data file not found.");
                return;
            }

            // Part 1
            int safeReportsCount = CountSafeReports(filePath, withDampener: false);
            Console.WriteLine($"Day 2 - Part 1: Number of safe reports: {safeReportsCount}");

            // Part 2
            int safeReportsWithDampener = CountSafeReports(filePath, withDampener: true);
            Console.WriteLine($"Day 2 - Part 2: Number of safe reports with Dampener: {safeReportsWithDampener}");
        }
    }
}

File: python/Day2.py
from typing import List

def is_safe_report(report: List[int]) -> bool:
    """
    Determine if a report is "safe". A report is safe if the difference between consecutive
    elements is either always between -3 and -1 (inclusive) or always between 1 and 3 (inclusive).

    Parameters:
        report (List[int]): A list of integers representing the report.

    Returns:
        bool: True if the report is safe, False otherwise.
    """
    differences = [b - a for a, b in zip(report, report[1:])]
    if all(-3 <= d <= -1 for d in differences):
        return True
    if all(1 <= d <= 3 for d in differences):
        return True
    return False

def is_safe_with_dampener(report: List[int]) -> bool:
    """
    Check if a report can be made safe by removing exactly one level (one element from the list).
    If the report is already safe, return True.
    Otherwise, try removing each element and check if it becomes safe.

    Parameters:
        report (List[int]): The report list.

    Returns:
        bool: True if the report can be safe after removing zero or one element, False otherwise.
    """
    if is_safe_report(report):
        return True

    for i in range(len(report)):
        modified_report = report[:i] + report[i+1:]
        if is_safe_report(modified_report):
            return True

    return False

def count_safe_reports(filename: str, with_dampener: bool = False) -> int:
    """
    Count the number of safe reports from a file.

    If with_dampener is False, checks if reports are safe as-is.
    If with_dampener is True, checks if reports can be safe with the dampener applied.

    Parameters:
        filename (str): Path to the file containing reports.
        with_dampener (bool): Whether to check the dampener scenario.

    Returns:
        int: Number of safe reports.
    """
    safe_count = 0
    with open(filename, 'r') as file:
        for line in file:
            report = list(map(int, line.split()))
            if with_dampener:
                if is_safe_with_dampener(report):
                    safe_count += 1
            else:
                if is_safe_report(report):
                    safe_count += 1
    return safe_count

def main():
    """Execute Day 2 puzzle solution."""
    file_path = '../data/Day2.txt'

    # Part 1
    safe_reports_count = count_safe_reports(file_path)
    print(f"Day 2 - Part 1: Number of safe reports: {safe_reports_count}")

    # Part 2
    safe_reports_with_dampener = count_safe_reports(file_path, with_dampener=True)
    print(f"Day 2 - Part 2: Number of safe reports with Dampener: {safe_reports_with_dampener}")

if __name__ == "__main__":
    main()
